src/ppom-controller.ts

```diff
@@ -0,0 +1,387 @@
+   */
+  #isOutOfDate(): boolean {
+    return Date.now() - this.state.lastFetched >= this.config.refreshInterval;
+  }
+
+  /**
+   * Update the PPOM configuration.
+   * This function will fetch the latest version info when needed, and update the PPOM storage.
+   */
+  async updatePPOM() {
+    if (this.#ppom) {
+      this.#ppom.free();
+      this.#ppom = undefined;
+    }
+
+    if (this.#isOutOfDate()) {
+      await this.#updateVersionInfo();
+    }
+
+    this.update({ lastChainId: this.state.newChainId });
+
+    const storageMetadata = await this.#storage.syncMetadata(
+      this.state.versionInfo,
+    );
+    const newFiles = await this.#getNewFiles(
+      this.state.newChainId,
+      storageMetadata,
+    );
+
+    for (const file of newFiles) {
+      await this.#storage.writeFile(file);
+    }
+  }
+
+  /**
+   * Returns an array of new files that should be downloaded and saved to storage.
+   *
+   * @param chainId - The chain ID to check for files.
+   * @param storageMetadata - An array of file metadata objects already in storage.
+   * @returns A promise that resolves to an array of new files to download and save to storage.
+   */
+  async #getNewFiles(chainId: string, storageMetadata: any[]): Promise<any[]> {
+    const newFiles: any[] = [];
+
+    for (const fileVersionInfo of this.state.versionInfo) {
+      //  download all files for the current chain + generally required files.
+      if (fileVersionInfo.chainId && fileVersionInfo.chainId !== chainId) {
+        continue;
+      }
+
+      // check if file is already in storage
+      if (
+        storageMetadata.find(
+          (file) =>
+            file.name === fileVersionInfo.name &&
+            file.chainId === fileVersionInfo.chainId &&
+            file.version === fileVersionInfo.version &&
+            file.checksum === fileVersionInfo.checksum,
+        )
+      ) {
+        continue;
+      }
+
+      const fileUrl = `${PPOM_CDN_BASE_URL}${fileVersionInfo.filePath}`;
+      const fileData = await this.#fetchBlob(fileUrl);
+      if (!fileData) {
+        throw new Error('Failed to fetch file data');
+      }
+
+      newFiles.push({
+        data: fileData,
+        ...fileVersionInfo,
+      });
+    }
+
+    return newFiles;
+  }
+
+  /*
+   * Fetch the version info from the PPOM cdn.
+   *  update the version info in state.
+   */
+  async #updateVersionInfo() {
+    const versionInfo = await this.#fetchVersionInfo(PPOM_VERSION_PATH);
+    if (!versionInfo) {
+      throw new Error('Failed to fetch version info');
+    }
+
+    this.update({
+      versionInfo,
+      lastFetched: Date.now(),
+    });
+  }
+
+  /**
+   * Conditionally update the ppom configuration.
+   *
+   * If the ppom configuration is out of date, this function will call `updatePPOM`
+   * to update the configuration.
+   */
+  async #maybeUpdatePPOM() {
+    if (await this.#shouldUpdate()) {
+      await this.updatePPOM();
+    }
+
+    this.update(this.state);
```

Comment:
Is this necessary? Does `updatePPOM` not already update the necessary state?

Reply:
I removed it